### PR TITLE
INTEL-MKL:  MKL primitive reuse for conv2d fwd op - refactoring per PR suggestion

### PR DESCRIPTION
**Pull request by gzmkl**
This PR updates the implementation of MKL primitive reuse for conv2d forward op, based on 
  1. Rasmus's initial PR review suggestion: moves private members of  MklConv2DFwdPrimitive into a structure. 
  2. Rename related implementation classes, for instance
       ConvFwdDimensions  -->  MklConvFwdParams  
  to make them more explicit/generic and to avoid any confusion with prefix "Mkl" since 
  they are under "tensorflow" namespace. 

I have temporarily close 4 related new MKL primitive reuse PRs (for conv bwd, relu, batchnorm, pooling)
since they are pending on this one.

### CODE REVIEW

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 1]
```diff
@@ -119,139 +125,174 @@ class Conv2DFwd : public DnnOp {
...
+    ConvFwdContext() :
+       src_fmt(memory::format::any), filter_fmt(memory::format::any),
+       src_mem(nullptr), filter_mem(nullptr), bias_mem(nullptr),
+       dst_mem(nullptr), fwd_desc(nullptr),
+       src_md(nullptr), filter_md(nullptr), bias_md(nullptr),
+       fwd_pd(nullptr), conv_fwd(nullptr), fwd_stream(nullptr) {
+    }
+  } context_;
+
+  engine cpu_engine_ = engine(engine::cpu, 0);
```
Comment (rmlarsen) [inline tensorflow/core/kernels/mkl_conv_ops.cc]: Move the initialization to initializer list in constructor instead?

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 2]
```diff
@@ -119,139 +125,174 @@ class Conv2DFwd : public DnnOp {
...
-  std::shared_ptr<mkldnn::primitive> conv_fwd_;
+  std::shared_ptr<mkldnn::convolution_forward::primitive_desc>
+  GetPrimitiveDesc() const {
+    return context_.fwd_pd;
+  }
 
  private:
-  void Setup(const ConvFwdDimensions& convFwdDims) {
+  // Primitive reuse context for Conv2D Fwd op
+  struct ConvFwdContext {
```
Comment (rmlarsen) [inline tensorflow/core/kernels/mkl_conv_ops.cc]: Data members should be declared last. So just leave the type definition of ConvFwdContext here, and more the declaration of context_ and engine_ to the bottom of the class.

**Review by rmlarsen — COMMENTED** [review]
A few minor comments.

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 3]
```diff
@@ -82,35 +83,41 @@ struct ConvFwdDimensions {
 };
 
 template <typename T>
-class Conv2DFwd : public DnnOp {
+class MklConv2DFwdPrimitive: public MklPrimitive {
  public:
-  explicit Conv2DFwd(const ConvFwdDimensions& convFwdDims) {
-    fwd_stream_.reset(new stream(stream::kind::eager));
+  explicit MklConv2DFwdPrimitive(const MklConvFwdParams& convFwdDims) :
+    cpu_engine_(engine(engine::cpu, 0)) {
```
Comment (rmlarsen) [inline tensorflow/core/kernels/mkl_conv_ops.cc]: Perhaps  cpu_engine_(engine::cpu, 0) would be simple?

**Review by rmlarsen — DISMISSED** [review]

File: tensorflow/core/kernels/mkl_conv_ops.cc (hunk 1 above)
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done with code refactoring.

File: tensorflow/core/kernels/mkl_conv_ops.cc (hunk 2 above)
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done with code refactoring - move both "context_" and "cpu_engine_" data members to 
the bottom of class declaration.

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 4]
```diff
@@ -119,139 +125,174 @@ class Conv2DFwd : public DnnOp {
...
+    context_.dst_mem->set_data_handle(
+        static_cast<void*>(const_cast<T*>(dst_data)));
+    context_.fwd_stream->submit(context_.fwd_primitives);
+
+    // after execution, set data handle back
+    context_.src_mem->set_data_handle(DummyData);
+    context_.filter_mem->set_data_handle(DummyData);
+    context_.dst_mem->set_data_handle(DummyData);
 
     return;
```
[the remark this answers is not in the document]
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done by removing the "return" statement and the empty line above.

File: tensorflow/core/kernels/mkl_conv_ops.cc (hunk 3 above)
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Yes, I have done by (1) declaring it at bottom of class and (2) moving initialization to constructor's initialization list. Thanks!

Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Sorry, this commit is for the new code. Let me double check.

**Review by rmlarsen — APPROVED** [review]

**Comment by gzmkl** [comment]
Hi Ramesh, 
Thank you very much for your valuable suggestions. 
I will do similar code refactoring on the remaining 4 MKL primitive reuse PRs (conv2d bwd, relu, batchnorm and pooling) before re-opening them for your review.
--GZ

**Comment by gzmkl** [comment]
Hi Rasmus,
I am sorry, "Hi Ramesh" should be "Hi Rasmus" in my previous comment. I was confused 
with my team leader's name. 
GZ
